src/Utilities/TaskItem.cs

```diff
@@ -298,9 +298,14 @@ public void CopyMetadataTo(ITaskItem destinationItem)
             if (_metadata != null)
             {
                 // Avoid a copy if we can
-                if (destinationItem is TaskItem destinationAsTaskItem && destinationAsTaskItem.Metadata == null)
+                if (destinationItem is TaskItem destinationAsTaskItem && destinationAsTaskItem.Metadata?.Count < _metadata?.Count)
                 {
-                    destinationAsTaskItem.Metadata = _metadata.Clone(); // Copy on write!
+                    CopyOnWriteDictionary<string> copiedMetadata = _metadata.Clone(); // Copy on write!
+                    foreach (KeyValuePair<string, string> entry in destinationAsTaskItem.Metadata)
+                    {
+                        copiedMetadata[entry.Key] = entry.Value;
```

Comment:
Could you clarify the purpose of this change? Is this to optimistically optimize the case where you are copying metadata values that exactly match the destination (in value and count) - in this case the foreach here will not cause a "copy" of the dictionary because it will just be writing the same values? If that's the case, why not `destinationAsTaskItem.Metadata?.Count == _metadata?.Count`?

Reply:
There was previously an optimization for if destinationAsTaskItem.Metadata was null, i.e., its count was 0, but there are also cases in which its count is small compared to _metadata?.Count. This is trying to improve that case, where there are a few metadata defined for destinationAsTaskItem but not as many as for the source. In that case, it's better to modify a clone of the source rather than modifying the destination directly.